provider/negotiate.go

```diff
@@ -27,36 +28,35 @@ func (provider *Provider) NegotiateContract(contract *core.Contract) (*core.Cont
 	// Sign contract
 	provSig, err := core.SignContract(contract, provider.PrivateKey)
 	if err != nil {
-		return nil, errors.New("Error signing contract")
+		return nil, fmt.Errorf("Failed to sign contract. error: %s", err)
 	}
 	contract.ProviderSignature = provSig
 
-	// Add storage space to the renter
+	// if renter isn't in set, add a new entry
 	renter, exists := provider.renters[contract.RenterId]
 	if !exists {
-		renter = &RenterInfo{
-			Contracts: []*core.Contract{},
-			Blocks:    []*BlockInfo{},
-		}
+		renter = &RenterInfo{}
 		provider.renters[contract.RenterId] = renter
 	}
+	// Add storage space to the renter
 	renter.StorageReserved += contract.StorageSpace
-	renter.Contracts = append(renter.Contracts, contract)
-
-	provider.contracts = append(provider.contracts, contract)
-	provider.addActivity("contract", contract.StorageSpace)
 
-	err = provider.UpdateMeta()
+	err = provider.InsertContract(contract)
 	if err != nil {
-		return nil, fmt.Errorf("Error updating metaserver: %s", err)
+		return nil, fmt.Errorf("Failed to insert contract into DB. error: %s", err)
 	}
 
-	err = provider.saveSnapshot()
+	// activity updates are non-fatal errors
+	err = provider.addActivity("contract", contract.StorageSpace)
 	if err != nil {
+		fmt.Println("Failed to update activity for contract: ", err)
+	}
 
-		// TODO: Remove contract. I don't do this here
-		// since we need to move to an improved storage scheme anyways.
-		return nil, fmt.Errorf("Unable to save contract. Error: %s", err)
+	// this could potentially be non-fatal too
+	err = provider.UpdateMeta()
```

Comment:
This is just every successful contract creation. I'm not sure where else we would do this as otherwise it's only updated on config update (which really should not happen often).

Reply:
Here's an idea:

```
func (p *Provider) metaUpdateThread() {
   updateInterval := 5 * time.Minute
   ticker := time.NewTicker(updateInterval)
   for {
      switch {
         case <- tickerC:
             p.updateMeta()
         case <- p.doneCh:
             ticker.Stop()
             return
    }
}

func (p *Provider) StartBackgroundThreads() {
   go p.metaUpdateThread()
}

func (p *Provider) StopBackgroundThreads() {
   close(p.doneCh)
}
```

Reply:
Or a little more sophisticated: 

```
func (p *Provider) metaUpdateThread() {
   updateInterval := 5 * time.Minute
   ticker := time.NewTicker(updateInterval)
   pinfo := p.GetInfo()
   for {
      switch {
         case <- ticker.C:
             updatedPinfo := p.GetInfo()
             if *updatedPinfo != *pinfo {
                p.updateMeta()
             }
             pinfo = updatedPinfo
         case <- p.doneCh:
             ticker.Stop()
             return
        }
    }
}

func (p *Provider) StartBackgroundThreads() {
   go p.metaUpdateThread()
}

func (p *Provider) StopBackgroundThreads() {
   close(p.doneCh)
}
```